sycl/doc/extensions/SPIRV/SPV_INTEL_float_controls2.asciidoc

```diff
@@ -0,0 +1,265 @@
+
+Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products. If you are interested in using this feature in your software product, please let us know!
+
+
+Version
+-------
+
+[width="40%",cols="25,25"]
+|========================================
+| Last Modified Date | 2020-12-03
+| Revision           | 4
+|========================================
+
+Dependencies
+------------
+
+This extension is written against the SPIR-V Specification,
+Version 1.5, Revision 3, Unified
+
+This extension requires SPIR-V 1.0.
+
+Overview
+--------
+
+This extension adds new execution modes and decorations to control floating-point computations.
+This extension adds execution modes with round to positive infinity and round to negative infinity default rounding
+and execution modes specifying compliance with floating-point arithmetic (IEEE 754) standard.
+This extension adds new decorations that can be applied to a function to control floating-point computations inside the function.
+
+
+Extension Name
+--------------
+
+To use this extension within a SPIR-V module, the following
+*OpExtension* must be present in the module:
+
+----
+OpExtension "SPV_INTEL_float_controls2"
+----
+
+Modifications to the SPIR-V Specification, Version 1.5, Revision 3, Unified
+---------------------------------------------------------------------------
+
+Modify Section 2.16.1, Universal Validation Rules:
+
+Modify the Entry Point rules:
+
+Add *RoundingModeRTPINTEL* and *RoundingModeRTNINTEL* to the list of execution modes in the third rule.
+
+Add the following rule:
+
+- Each OpEntryPoint must not set more than one of the *FloatingPointModeALTINTEL* or *FloatingPointModeALTINTEL* execution modes for any given _Target Width_.
+
+Modify the Decoration rules, add the following rules:
+
+- If an *OpFunction* is decorated by more than one *FunctionRoundingModeINTEL* for any given _Target Width_ the _FP Rounding Mode_ operands must match.
+- If an *OpFunction* is decorated by more than one *FunctionDenormModeINTEL* for any given _Target Width_ the _FP Denorm Mode_ operands must match.
+- If an *OpFunction* is decorated by more than one *FunctionFloatingPointModeINTEL* for any given _Target Width_ the _FP Operation Mode_ operands must match.
+
+Modify Section 3.6, Execution Mode:
+
+Add the following rule to the description of *DenormPreserve* execution mode:
+
+If an instruction is inside of a function decorated with *FunctionDenormModeINTEL* with matching _Target Width_ operand, the denorm mode specified by *FunctionDenormModeINTEL* is applied and *DenormPreserve* is ignored.
+
+Add the following rule to the description of *DenormFlushToZero* execution mode:
+
+If an instruction is inside of a function decorated with *FunctionDenormModeINTEL* with matching _Target Width_ operand, the denorm mode specified by *FunctionDenormModeINTEL* is applied and *DenormFlushToZero* is ignored.
```

Comment:
It's fine if we defer this to a client API environment spec, but so we're thinking about it: How should these execution modes and decorations interact with program build options like `-cl-denorms-are-zero`?  Does the decoration or the build option "win"?

Reply:
Does `-cl-denorms-are-zero` build option cause that the corresponding **DenormPreserve** Execution Mode is present in the shader?

Reply:
The `-cl-denorms-are-zero` build option is be specified when compiling SPIR-V, not when generating SPIR-V, so I think we'd need to work out the interactions between the build option and the execution mode.  I don't think this has been an issue before now because most of the compute consumers haven't moved to SPIR-V 1.4 (and hence don't need to worry about **DenormPreserve**), but it's something we'll need to figure out eventually.

Do we have a recommendation how this should work?